subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2092 @@
+		return -EINVAL;
+	}
+
+	if (!(ag->hf_features & BT_HFP_HF_FEATURE_CODEC_NEG) ||
+	    !(ag->ag_features & BT_HFP_AG_FEATURE_CODEC_NEG)) {
+		return -EOPNOTSUPP;
+	}
+
+	while (error == 0) {
+		error = get_number(buf, &number);
+
+		if (error != 0) {
+			return error;
+		}
+
+		ret = is_char(buf, ',');
+		if (ret != true) {
+			if (true != is_char(buf, '\r')) {
+				return -EINVAL;
+			}
+		}
+
+		if (number < (sizeof(codec_ids) * 8)) {
+			codec_ids |= BIT(number);
+		}
+	}
+
+	ag->hf_codec_ids = codec_ids;
+
+	if (bt_ag && bt_ag->codec) {
+		bt_ag->codec(ag, ag->hf_codec_ids);
+	}
+
+	return 0;
+}
+
+static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err)
+{
+	int error;
+	bool inquiry_status = true;
+	bool ret;
+
+	*err = CME_ERROR_AG_FAILURE;
+
+	ret = is_char(buf, '=');
+	if (ret == true) {
+		inquiry_status = false;
+	}
+
+	ret = is_char(buf, '?');
+	if (ret != true) {
+		return -EINVAL;
+	}
+
+	ret = is_char(buf, '\r');
+	if (ret != true) {
+		return -EINVAL;
+	}
+
+	if (false == inquiry_status) {
+		error = hfp_ag_send_data(
+			ag, NULL, NULL,
+			"\r\n+CIND:(\"%s\",(%d%s%d)),(\"%s\",(%d%s%d)),(\"%s\",(%d%s%d)),(\"%s\",(%"
+			"d%s%d)),(\"%s\",(%d%s%d)),(\"%s\",(%d%s%d)),(\"%s\",(%d%s%d))\r\n",
+			ag_ind[BT_HFP_AG_SERVICE_IND].name, ag_ind[BT_HFP_AG_SERVICE_IND].min,
+			ag_ind[BT_HFP_AG_SERVICE_IND].connector, ag_ind[BT_HFP_AG_SERVICE_IND].max,
+			ag_ind[BT_HFP_AG_CALL_IND].name, ag_ind[BT_HFP_AG_CALL_IND].min,
+			ag_ind[BT_HFP_AG_CALL_IND].connector, ag_ind[BT_HFP_AG_CALL_IND].max,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].name, ag_ind[BT_HFP_AG_CALL_SETUP_IND].min,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].connector,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].max, ag_ind[BT_HFP_AG_CALL_HELD_IND].name,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].min,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].connector,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].max, ag_ind[BT_HFP_AG_SINGNAL_IND].name,
+			ag_ind[BT_HFP_AG_SINGNAL_IND].min, ag_ind[BT_HFP_AG_SINGNAL_IND].connector,
+			ag_ind[BT_HFP_AG_SINGNAL_IND].max, ag_ind[BT_HFP_AG_ROAM_IND].name,
+			ag_ind[BT_HFP_AG_ROAM_IND].min, ag_ind[BT_HFP_AG_ROAM_IND].connector,
+			ag_ind[BT_HFP_AG_ROAM_IND].max, ag_ind[BT_HFP_AG_BATTERY_IND].name,
+			ag_ind[BT_HFP_AG_BATTERY_IND].min, ag_ind[BT_HFP_AG_BATTERY_IND].connector,
+			ag_ind[BT_HFP_AG_BATTERY_IND].max);
+	} else {
+		error = hfp_ag_send_data(ag, NULL, NULL, "\r\n+CIND:%d,%d,%d,%d,%d,%d,%d\r\n",
+					 ag->indicator_value[BT_HFP_AG_SERVICE_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_SETUP_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_HELD_IND],
+					 ag->indicator_value[BT_HFP_AG_SINGNAL_IND],
+					 ag->indicator_value[BT_HFP_AG_ROAM_IND],
+					 ag->indicator_value[BT_HFP_AG_BATTERY_IND]);
+	}
+
+	return error;
+}
+
+static void bt_hfp_ag_set_in_band_ring(struct bt_hfp_ag *ag, void *user_data)
+{
+	if (0 != (ag->ag_features & BT_HFP_AG_FEATURE_INBAND_RINGTONE)) {
```

Comment:
Please don't use this kind of constructions with the constant on the left side, since it's not as easy for a human to parse. If you end up having lots of places testing for the content of `ag->ag_features` you could even create helper macros to make the code more readable, something like:
```c
#define AG_FEATURE_SUPPORTED(ag, feat) (((ag)->ag_features & feat) != 0)
#define HF_FEATURE_SUPPORTED(ag, feat) (((ag)->hf_features & feat) != 0)
```